pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -806,11 +822,8 @@ protected void doRemoveSegment(IndexSegment segment) {
     long startTimeMs = System.currentTimeMillis();
 
     MutableRoaringBitmap validDocIds =
-        segment.getValidDocIds() != null ? segment.getValidDocIds().getMutableRoaringBitmap() : null;
-    if (validDocIds == null || validDocIds.isEmpty()) {
-      _logger.info("Skip removing segment without valid docs: {}", segmentName);
-      return;
-    }
+        segment.getValidDocIds() != null ? segment.getValidDocIds().getMutableRoaringBitmap()
+            : new MutableRoaringBitmap();
```

Comment:
We should early terminate here

Reply:
Not in all scenarios, we'd want to terminate early here. Like sometimes validDocID might not be there at all in a segment but we had still want to iterate the segments to find distinct primary-keys.

Reply:
I don't follow. When there is no valid docs, why would we iterate over an empty bitmap? It will be no-op but wasting CPU cycles

Reply:
What I meant is in the base class, early termination might be an issue since in another extension, we might want to iterate through the segment and not just valid doc ids.

Reply:
Moved this early termination to the function `removeSegment(segment, validDocIds` so the optimisation is preserved.

Reply:
Can you elaborate more on this:

> What I meant is in the base class, early termination might be an issue since in another extension, we might want to iterate through the segment and not just valid doc ids.

Trying to see what functionality you are seeking for.
Even after moving the early termination to `removeSegment()`, there will be extra overhead following `removeSegment()` and logging changes. More importantly, the current contract for `removeSegment()` is that the segment contains valid docs (it is also invoked when replacing segment). In order to handle segment removal differently, you can override `doRemoveSegment()` instead

Reply:
>  In order to handle segment removal differently, you can override doRemoveSegment() instead

Yes that's what I realised today that we can override `doRemoveSegment()` as well in our implementation. I was just trying to reuse code as much as possible and move everything to base class so that we don't add additional developer overhead to do changes in multiple places (in base class and in extended class). 

But if you think this refactoring is causing major un-optimisations to segment-removal flow, I am fine with closing this PR then and will override `doRemoveSegment()`.

Reply:
The changes to make a primary key iterator is okay if that helps with the testing logic.
I'd suggest reverting the changes for `doRemoveSegment()` because currently the implementation is assuming that only valid docs need to be removed. If you need different handling, then overriding `doRemoveSegment()` is the correct thing to do. My concern is mostly on the high level abstraction/assumption held for each function. Changing that can lead to other problems in the long run.

Reply:
Hmm. So right now, `doRemoveSegment()` does the following:

> 1. log start of removal segment
> 2. get valid doc id for segment
> 3. validate valid doc id 
> 4. removeSegment(segment, validDocID)
> 5. logs and update metrics

I was trying to move the validate to `removeSegment(segment, validDocID)` so that I don't need to reimplement the other 3 steps in my implementation of `doRemoveSegment(segment)` just to avoid the validation step. 
This will also couple validation of valid doc id in `removeSegment(segment, validDocID)` method which seems right.

Reply:
Notice that `removeSegment(segment, validDocIds)` is also used in segment replacement (after validation). The whole method of `doRemoveSegment()` is just about 10 lines, so let's not change it and simply override if necessary.

Reply:
Updated!